src/irmin/node_intf.ml

```diff
@@ -192,24 +254,95 @@ end
 
 module type Sigs = sig
   module type S = S
-  module type Maker = Maker
 
-  module Make : Maker
-  (** [Make] provides a simple node implementation, parameterized by the
-      contents and notes keys [K], paths [P] and metadata [M]. *)
+  (** [Make] provides a simple node implementation, parameterized by hash, path
+      and metadata implementations. The contents and node values are addressed
+      directly by their hash. *)
+  module Make
+      (Hash : Hash.S) (Path : sig
+        type step [@@deriving irmin]
+      end)
+      (Metadata : Metadata.S) :
+    S
+      with type hash = Hash.t
+       and type metadata = Metadata.t
+       and type step = Path.step
+
+  (** [Generic_key] generalises the concept of "node" to one that supports
+      object keys that are not strictly equal to hashes. *)
+  module Generic_key : sig
+    module type S = S_generic_key
+    module type Maker = Maker_generic_key
+
+    module Make : Maker
+
+    module Store
+        (C : Contents.Store)
+        (S : Indexable.S)
+        (H : Hash.S with type t = S.hash)
+        (V : S
+               with type t = S.value
+                and type hash = H.t
+                and type contents_key = C.key
+                and type node_key = S.key)
+        (M : Metadata.S with type t = V.metadata)
+        (P : Path.S with type step = V.step) :
+      Store
+        with type 'a t = 'a C.t * 'a S.t
+         and type key = S.key
+         and type hash = S.hash
+         and type value = S.value
+         and module Path = P
+         and module Metadata = M
+         and module Val = V
+  end
 
   (** v1 serialisation *)
-  module V1 (N : S with type step = string) : sig
+  module V1 (N : Generic_key.S with type step = string) : sig
     include
-      S
-        with type hash = N.hash
+      Generic_key.S
+        with type contents_key = N.contents_key
+         and type node_key = N.node_key
          and type step = N.step
          and type metadata = N.metadata
 
     val import : N.t -> t
     val export : t -> N.t
   end
 
+  module Portable : sig
```

Comment:
I am not totally sold on the terminology here (not sure if `Portable` is exactly what we want here) but I have no better suggestions yet 😁

Reply:
Agreed. I also considered `Hash_preimage` and `Hash_keyed`; not sure if either of those are better.

Reply:
`Node.Hashable` ? Or `Node.with_internal_keys_replaced_by_hashes` 📦 

Or maybe just call it a "normal" or "canonical" form? e.g. `Node.Canonical` ?

Reply:
I like `Hashable`

Reply:
I don't much like `Hashable`, because the regular `Node` type is also hashable. I think the thing that distinguishes this from a regular node is that it's transmissable between stores (hence portable), but let's resolve this with a Slack poll. (I hear democracy always produces optimal results.)